### PR TITLE
AuthRequiredScope cannot access PresenterProvider

### PR DESCRIPTION
**Pull request by viceversus**
Please do not merge. This is very early stages of WIP. I ran into a problem where the `retained.cast` for AuthRequiredScope was failing because `Presenter.PresenterProvider` was only scoped to `AuthOptional`. I'm not that familiar with Anvil so I wasn't able to figure out a better solution than to dupe the Presenter. We're explicitly passing `AuthOptionalScope` or `AuthRequiredScope`, so I can't put it into a higher level scope. It probably doesn't belong there anyway. Maybe this could go in `AuthAwareScope`? I don't know the details of the domain well enough to make that call.

### CODE REVIEW

File: skeleton/src/main/java/com/dropbox/kaiken/skeleton/scoping/SingleIn.kt [hunk 1]
```diff
@@ -120,6 +121,7 @@ interface AuthOptionalScreenComponent : Injector {
     parentScope = UserScope::class
 )
 @SingleIn(AuthRequiredScreenScope::class)
+@MergeComponent(AuthOptionalScreenScope::class)
```
Comment (digitalbuddha) [inline skeleton/src/main/java/com/dropbox/kaiken/skeleton/scoping/SingleIn.kt:124]: AuthRequiredScreenScope

we want this component to have all the objects that do @ContributesBinding(AuthRequiredScreenScope)

This way we can separate what goes into the AuthOptionalScreenComponent from the AuthRequiredScreenComponent

File: sample/sample-with-di/helloworldfeature/src/main/java/com/dropbox/kaiken/sample/advancedsample/helloworldfeature/ComposeBindings.kt [hunk 2]
```diff
@@ -149,6 +152,25 @@ inline fun <reified T : BasePresenter, reified V : Injector> retainComponentAndS
     }
 }
 
+@Composable
+inline fun <reified T : BasePresenter, reified V : Injector> retainAuthedComponentAndSetContent(
```
Comment (digitalbuddha) [inline sample/sample-with-di/helloworldfeature/src/main/java/com/dropbox/kaiken/sample/advancedsample/helloworldfeature/ComposeBindings.kt:156]: this looks great good catch!.

File: skeleton/src/main/java/com/dropbox/kaiken/skeleton/scoping/SingleIn.kt (hunk 1 above)
Comment (digitalbuddha) [reply skeleton/src/main/java/com/dropbox/kaiken/skeleton/scoping/SingleIn.kt:124]: actually im not sure this line is needed at all. ContriutsSubcomponent is sort of like a merge component

File: sample/sample-with-di/helloworldfeature/src/main/java/com/dropbox/kaiken/sample/advancedsample/helloworldfeature/HomeActivity.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import androidx.navigation.compose.NavHost
+import androidx.navigation.compose.currentBackStackEntryAsState
+import androidx.navigation.compose.rememberNavController
+import com.dropbox.kaiken.sample_with_di.helloworldfeature.R
+
+
+class HomeActivity: AuthRequiredComposeActivity() {
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        if (finishIfInvalidAuth()) return
```
Comment (digitalbuddha) [inline sample/sample-with-di/helloworldfeature/src/main/java/com/dropbox/kaiken/sample/advancedsample/helloworldfeature/HomeActivity.kt:28]: now that we have AuthRequiredComposeActivity as an abstract class I think we should put this if statement in that class instead

File: skeleton/src/main/java/com/dropbox/kaiken/skeleton/scoping/SingleIn.kt (hunk 1 above)
Comment (viceversus) [reply skeleton/src/main/java/com/dropbox/kaiken/skeleton/scoping/SingleIn.kt:124]: Oops, this was me trying to fix something and forgetting to take it out 😅

**Pull request closed without merging**